### DISCUSSION TITLE
Singularity HDF5 Point Obs Readin

### DISCUSSION THREAD

**Discussion by justintsu** [dtcenter/METplus#2391] [Configuration] [answered]
Hello,

I am trying to get point observations (radiosondes for now) in an HDF5 file read into point_stat using embedding. I've written the script test_h5_embed.py which reads in the hdf5 file and rearranges the incoming raobs in a dataframe whose indices match what I read on the MET documentation 

_df_master = pd.DataFrame(columns=['typ','sid','vld','lat','lon','elv','var',\
                         'lvl','hgt','qc','obs'])_

Finally, the dataframe values are extracted and converted to a list and stored in a variable named met_point_obs (although I've also tried storing it in varialbes named point_obs_data and point_data, neither worked).  The script is invoked with arguments specifying what particular observation to store in 'obs'.  I've placed a temperature file  along with this script attached to this discussion.  The script is invoked like:

_test_h5_embed.py NAF_raob_2023033000.h5 tmp_

With the correct environmental variables set, I run:

_singularity exec /app/projects/NEPTUNE/software/singularity/met_v11.0.1.simg  point_stat 'PYTHON_NUMPY' 'PYTHON_NUMPY' PointStat2023033000NEST1a0121x0121  -v 5  -outdir ./out/point_stat/radiosonde_

and get this non-descript error:

**ERROR  :
ERROR  : process_obs_file() -> trouble parsing the python command PYTHON_NUMPY.
ERROR  :**

[Code_Config.zip](https://github.com/dtcenter/METplus/files/13169958/Code_Config.zip)

(See below comment for the observation file)

**Comment by justintsu** [comment]
Here is the observation file split into two files.  You can extract the full data file by running (post unzip)

cat NAF_raob_2023033000.tar.gz.a* | tar xzvf -

[NAF_raob_2023033000.tar.gz.ab.zip](https://github.com/dtcenter/METplus/files/13170026/NAF_raob_2023033000.tar.gz.ab.zip)
[NAF_raob_2023033000.tar.gz.aa.zip](https://github.com/dtcenter/METplus/files/13170027/NAF_raob_2023033000.tar.gz.aa.zip)

**Comment by DanielAdriaansen** [comment]
Greetings @justintsu. I see that you are having trouble with Python embedding for point observations using Point Stat. Thank you for providing your script, data files, Point Stat config file, and command.

Prior to MET v11.0.0, Python Embedding for point observations was not well supported in tools like Point Stat, which require a special format of point observations typically provided by the PB2NC tool. In MET v11.0.0 and v11.0.1 (the version you are using), in order to use Python embedding for point observations with Point Stat you'll need to call a special reformatter on your `point_data` variable in your Python embedding script so that it conforms to the format Point Stat expects. This was a temporary bridge solution to the permanent fix which is included in MET v11.1.0+, where this is handled automatically by MET. So just be aware the suggestion here only applies to MET v11.0.0 and 11.0.1.

If you cannot upgrade to the latest version of MET (v11.1.0), I would recommend the following changes to your Python script `test_h5_embed.py`:
1) On line 10, add: `from met_point_obs import convert_point_data`
2) Comment out line 226
3) Uncomment line 227
4) On line 228, add: `met_point_data = convert_point_data(point_data)`

Could you please try that and let us know how it works?

**Reply by justintsu** [reply]
Thank you for the reply,

I will try that but is there a way to install the met_point_obs module? Pip can't seem to find it.

**Reply by DanielAdriaansen** [reply]
The `met_point_obs` module is part of the MET v11.0.1 installation, and when you invoke your Python script with Point Stat (it may not work running by hand), it should find it.

If it does not work automatically or cannot find it, please try copying the file to the working directory where your Python script is. The file can be found at:
https://github.com/dtcenter/MET/blob/v11.0.1/scripts/python/met_point_obs.py

**Reply by justintsu** [reply]
thank you, I copied the file as a backup for my own reference and placed it (just in case) in the same directory as I am invoking test_h5_embed.py in.  Got the same error as in the original post

**Reply by DanielAdriaansen** [reply]
Great. Let's work on your Point Stat command.

Per the documentation for MET v11.0.1 here:
https://met.readthedocs.io/en/v11.0.1/Users_Guide/appendixF.html#python-embedding-for-point-observations

The `obs_file` command line argument for Point Stat needs to be 'PYTHON_NUMPY=/path/to/python/script.py arg1 arg2'. This differs from Python embedding for forecast data, where the Python embedding information is contained in the `name` entry of the `fcst` dictionary in the Point Stat config file.

Looking at your config file, I see you're placing all forecast and obs pairs in the file. For testing, let's reduce to a single fcst/obs file pair and scale up from there:
```
fcst = {
     field = [
{name = "/p/home/fnmocalf/user_dirs/tsu/uv/nrl/veriftools/model/coamps/read_COAMPS_general.py /p/work1/fnmocalf/user_dirs/tsu/data/ARCTIC/airtmp_pre_000100_000000_1a0121x0121_2023033000_00000000_fcstfld ";}
     ];
}
obs = {
    field = [
{name="";desc ="ARCTIC_Nest1a0121x0121";level=["P100.0-100.0"];}
    ];
}
```

Then, your Point Stat command would look like this:
```
singularity exec /app/projects/NEPTUNE/software/singularity/met_v11.0.1.simg point_stat 'PYTHON_NUMPY' 'PYTHON_NUMPY=/p/home/fnmocalf/user_dirs/tsu/uv/scripts/COAMPS/test_h5_embed.py /p/work1/fnmocalf/user_dirs/tsu/data/obs/NAF_raob_2023033000.h5 tmp' PointStat2023033000NEST1a0121x0121 -v 5 -outdir ./out/point_stat/radiosonde
```

See if you can get it to work for this single fcst/obs pair, and sorry I didn't notice this first!

**Reply by justintsu** [reply]
Thanks for the quick responses, Daniel.

Ive modified my wrapper script to simplify the number of varialbes and made the suggested changes.  I am noticing the PYTHON_NUMPY segments are fairly finnicky with the placement of single quotes versus double quotes.  If I wanted to parse out environment variables in the PYTHON_NUMPY statement, how would I do that? 

Writing:

      singularity exec $sing_dir/met_v11.0.1.simg \
       point_stat \
       'PYTHON_NUMPY' \
       **'PYTHON_NUMPY=${READ_OBS} ${OB_DIR}/${OBFILE} ${RAOB_NAME}' \**
       ${CONFIG} \
       -v 5 \
       -outdir ${PSOUT}

seemed to not parse out the variables.  Although, point_stat did attempt to run test_embed_h5.py, albeit unsuccessfully with the following error:

DEBUG 4:        /p/home/fnmocalf/user_dirs/tsu/.local/external/anaconda3/bin/python /app/projects/NEPTUNE/database/MET/met-11.0.1_share/wrappers/write_tmp_point_nc.py /tmp/tmp_met_nc_34729_0 ${READ_OBS} ${OB_DIR}/${OBFILE} ${RAOB_NAME}
Traceback (most recent call last):
  File "/app/projects/NEPTUNE/database/MET/met-11.0.1_share/wrappers/write_tmp_point_nc.py", line 23, in <module>
    from met_point_obs import met_point_obs
ModuleNotFoundError: No module named 'met_point_obs'
ERROR  :
ERROR  : tmp_nc_point_obs() -> command "/p/home/fnmocalf/user_dirs/tsu/.local/external/anaconda3/bin/python /app/projects/NEPTUNE/database/MET/met-11.0.1_share/wrappers/write_tmp_point_nc.py /tmp/tmp_met_nc_34729_0 ${READ_OBS} ${OB_DIR}/${OBFILE} ${RAOB_NAME}" failed ... status = 256
ERROR  :

It looks as if I might not have met_point_obs.py in my current build of MET or it might not exist in my singularity.  If so, should I add the path leading to met_point_obs.py to my PYTHONPATH?

Writing:

      singularity exec $sing_dir/met_v11.0.1.simg \
       point_stat \
       'PYTHON_NUMPY' \
       **"'PYTHON_NUMPY=${READ_OBS} ${OB_DIR}/${OBFILE} ${RAOB_NAME}'" \**
       ${CONFIG} \
       -v 5 \
       -outdir ${PSOUT}

seems to not process the PYTHON_NUMPY statement correctly, even though the environmental variables are parsed correctly:

WARNING:
WARNING: process_obs_file() -> can't open observation netCDF file: 'PYTHON_NUMPY=/p/home/fnmocalf/user_dirs/tsu/uv/scripts/COAMPS/test_h5_embed.py /p/work1/fnmocalf/user_dirs/tsu/data/obs/NAF_raob_2023033000.h5 tmp'

**Reply by DanielAdriaansen** [reply]
No problem!

I think before we address environment variables we need to prove that it can successfully run for a single time, with a command where you put in the values for the environment variables by hand. This is partially because I don't readily know the answer and partially because environment variable substitution and quotation marks on the command line can be different from system to system and Linux shell to Linux shell.

Can you try a command by hand (with no environment variables) and report what happens?

**Reply by justintsu** [reply]
Ok, sounds like a plan.

I appened the pathway pointing to met_point_obs.py to my PYTHONPATH and exported it.  then ran singularity from my command line. Got this error:

/met-11.0.1_share/wrappers/write_tmp_point_nc.py /tmp/tmp_met_nc_100429_0 /p/home/fnmocalf/user_dirs/tsu/uv/scripts/COAMPS/test_h5_embed /p/work1/fnmocalf/user_dirs/tsu/data/obs/NAF_raob_2023033000.h5 tmp
Traceback (most recent call last):
  File "/app/projects/NEPTUNE/database/MET/met-11.0.1_share/wrappers/write_tmp_point_nc.py", line 24, in <module>
    from met_point_obs_nc import nc_point_obs
ModuleNotFoundError: No module named 'met_point_obs_nc'
ERROR  :
ERROR  : tmp_nc_point_obs() -> command "/p/home/fnmocalf/user_dirs/tsu/.local/external/anaconda3/bin/python /app/projects/NEPTUNE/database/MET/met-11.0.1_share/wrappers/write_tmp_point_nc.py /tmp/tmp_met_nc_100429_0 /p/home/fnmocalf/user_dirs/tsu/uv/scripts/COAMPS/test_h5_embed /p/work1/fnmocalf/user_dirs/tsu/data/obs/NAF_raob_2023033000.h5 tmp" failed ... status = 256
ERROR  :

**Reply by DanielAdriaansen** [reply]
My apologies, I meant to include the command to try. Please try this command:
```
singularity exec /app/projects/NEPTUNE/software/singularity/met_v11.0.1.simg point_stat 'PYTHON_NUMPY' 'PYTHON_NUMPY=/p/home/fnmocalf/user_dirs/tsu/uv/scripts/COAMPS/test_h5_embed.py /p/work1/fnmocalf/user_dirs/tsu/data/obs/NAF_raob_2023033000.h5 tmp' PointStat2023033000NEST1a0121x0121 -v 5 -outdir ./out/point_stat/radiosonde
```

**Reply by justintsu** [reply]
My shell restarted so my PYTHONPATH reset.  Running the command you posted above without specifying PYTHONPATH to include met_point_obs.py gives me the same error as before:

Traceback (most recent call last):
  File "/app/projects/NEPTUNE/database/MET/met-11.0.1_share/wrappers/write_tmp_point_nc.py", line 23, in <module>
    from met_point_obs import met_point_obs
ModuleNotFoundError: No module named 'met_point_obs'
ERROR  :
ERROR  : tmp_nc_point_obs() -> command "/p/home/fnmocalf/user_dirs/tsu/.local/external/anaconda3/bin/python /app/projects/NEPTUNE/database/MET/met-11.0.1_share/wrappers/write_tmp_point_nc.py /tmp/tmp_met_nc_8614_0 /p/home/fnmocalf/user_dirs/tsu/uv/scripts/COAMPS/test_h5_embed /p/work1/fnmocalf/user_dirs/tsu/data/obs/NAF_raob_2023033000.h5 tmp" failed ... status = 256
ERROR  :

**Reply by DanielAdriaansen** [reply]
Can you confirm if this file exists:
`/app/projects/NEPTUNE/database/MET/met-11.0.1_share/met/python/met_point_obs.py`

Are you setting any environment variables prior to running the command?

**Reply by justintsu** [reply]
That file does not exist.  I have a /app/projects/NEPTUNE/database/MET/met-11.0.1_share/python subdirectory but it is empty.  I copied the file you sent me in the link from a previous post into my current working directory, since I seem to not have any python files in my met-11.0.1_share directory.

I set the environment variables that are in my config file prior to running singularity.  I also set MET_BASE, MET_PYTHON_EXE, AND SINGULARITY_BINDPATH.

MET_BASE is /app/projects/NEPTUNE/database/MET/met-11.0.1_share/python
MET_PYTHON_EXE points to my own build of python
SINGULARITY_BINDPATH is a list of directories needed for my singularity build

**Reply by DanielAdriaansen** [reply]
Can you try unsetting `MET_BASE`, by either logging out and back in to your shell or manually unsetting the value, and try this command again:
```
singularity exec /app/projects/NEPTUNE/software/singularity/met_v11.0.1.simg point_stat 'PYTHON_NUMPY' 'PYTHON_NUMPY=/p/home/fnmocalf/user_dirs/tsu/uv/scripts/COAMPS/test_h5_embed.py /p/work1/fnmocalf/user_dirs/tsu/data/obs/NAF_raob_2023033000.h5 tmp' PointStat2023033000NEST1a0121x0121 -v 5 -outdir ./out/point_stat/radiosonde
```

Are you setting `MET_BASE` to try and get this to work? Or are you setting `MET_BASE` because you're using a MET installation has had it's shared files moved somewhere else?

**Reply by justintsu** [reply]
looks like unsetting MET_BASE is working, my observations seemed to have made it into point_stat however all of them are being rejected due to obs var name:
DEBUG 2: Searching 2893618 observations from 624 messages.
DEBUG 2:
DEBUG 2: --------------------------------------------------------------------------------
DEBUG 2:
WARNING:
WARNING: check_hdr_str() -> output header column OBS_VAR is empty!
WARNING:
DEBUG 2: Processing airtmp/pre_000100 versus /P100, for observation type radiosonde, over region FULL, for inter             polation method NEAREST(1), using 0 matched pairs.
DEBUG 2: Number of matched pairs   = 0
DEBUG 2: Observations processed    = 2893618
DEBUG 2: Rejected: station id      = 0
DEBUG 2: Rejected: obs var name    = 2893618
DEBUG 2: Rejected: valid time      = 0
DEBUG 2: Rejected: bad obs value   = 0
DEBUG 2: Rejected: off the grid    = 0
DEBUG 2: Rejected: topography      = 0
DEBUG 2: Rejected: level mismatch  = 0
DEBUG 2: Rejected: quality marker  = 0
DEBUG 2: Rejected: message type    = 0
DEBUG 2: Rejected: masking region  = 0
DEBUG 2: Rejected: bad fcst value  = 0
DEBUG 2: Rejected: bad climo mean  = 0
DEBUG 2: Rejected: bad climo stdev = 0
DEBUG 2: Rejected: mpr filter      = 0
DEBUG 2: Rejected: duplicates      = 0



Yes, I was setting MET_BASE because it's shared filed were moved elsewhere (somewhere separate from the singularity package)

**Reply by DanielAdriaansen** [reply]
OK. If all of the shared files at `MET_BASE` were moved, then setting your `MET_BASE` should not produce this error. Are you sure you are using the correct location for `MET_BASE`?

Please let me know if you need guidance setting `MET_BASE`, but if not I'd recommend leaving it unset.

Can you modify the `obs` entry in this file `PointStat2023033000NEST1a0121x0121` to look like this:
```
obs = {
    field = [
{name="";desc ="ARCTIC_Nest1a0121x0121";}
    ];
}
```
The `level` entry doesn't need to be in there, as it should be provided in your Python script in the `lvl` column of your DataFrame, which it looks like you are already doing:
```
 lvl = data[typ]['pres'][:]
```

Run the command again after and see if matches are made.

**Reply by justintsu** [reply]
I am not exactly sure why MET_BASE was set to what it was, I was provided this wrapper script with it pre defined and decided not to touch it.  Removing MET_BASE seems to not break anything though so I will keep it out.  

I've modified the config file and reran and got this error:

DEBUG 4: VarInfoFactory::new_var_info() -> created new VarInfo object of type "FileType_Python_Numpy".
WARNING:
WARNING: RegridInfo::validate() -> Resetting the regridding width from 4 to 2 for regridding method "BILIN".
WARNING:
WARNING:
WARNING: RegridInfo::validate() -> Resetting the regridding width from 4 to 2 for regridding method "BILIN".
WARNING:
ERROR  :
ERROR  : VarInfo::set_level_info_grib() - either level or GRIB_lvl_typ, GRIB_lvl_val1 and GRIB_lvl_val2 (if necessary) must be specified in field information
ERROR  :

**Reply by DanielAdriaansen** [reply]
You are attempting to regrid your forecast data to an observation grid, but your observations are not gridded:
```
////////////////////////////////////////////////////////////////////////////////

//
// Verification grid
// May be set separately in each "field" entry
//
regrid = {
   to_grid    = OBS;
   method     = BILIN;
   width      = 4;
   vld_thresh = 0;
   shape      = SQUARE;
}
```

Try setting `to_grid = NONE;` instead. This may not be the only problem though. Could you also share this script:
/p/home/fnmocalf/user_dirs/tsu/uv/nrl/veriftools/model/coamps/read_COAMPS_general.py

**Comment by justintsu** [comment]
[read_COAMPS_general.zip](https://github.com/dtcenter/METplus/files/13210234/read_COAMPS_general.zip)
Here is that file. I'll make the to_grid change and retry

**Reply by justintsu** [reply]
I tried point_stat with the to_grid setting in regrid set to NONE and got the same error as the previous post

**Reply by DanielAdriaansen** [reply]
OK, thank you. I re-read through our old conversations, and I think I made an incorrect suggestion.

Can you modify your `obs` entry in your config file to look like this:
```
obs = {
    field = [
{name="tmp";desc ="ARCTIC_Nest1a0121x0121";level=["P100.0-100.0"];}
    ];
}
```

Please re-add the `level` information, and modify the `name` entry to be the variable name in the `var` column in your DataFrame, which I think is `tmp` for this test. See if this helps.

Sorry for the earlier confusion.

**Reply by justintsu** [reply]
no worries, it looks like it is working now.  It also looked like I was mixing up the name of hte variable with the name of the HDF5 structure.  Thank you!

**Reply by justintsu** [reply]
Even though that run was successful, it took about 9 minutes to complete even though I was verifying one variable (tmp) at one pressure level (100mb) and at one tau.  I will be verifying 6 variables at 12 hour taus from 0-48 at multiple datetime groups and all significant pressure levels, so I am imagining this to be a VERY lengthy point_stat run.  The long length of this run is likely due to the HDF5 read in which contains millions of raobs.  This leads me to two questions

1. When running point_stat with observation python embedding and multiple observations lines in the obs block in the config file, will point_stat attempt to execute the python embedding once or as many times as there are lines in the obs block?
2. Since the observation embedding is performed in the invocation of point_stat at the command line and since my observation processing python routine is only able to read in one variable into memory at a time, does this mean that I will need to run point_stat separately for each variable and hence point_stat will be reading in the same raob file once per variable?

**Reply by justintsu** [reply]
it seems like I should be able to modify my dataframe to have all the variable types in column 'var', and just run test_h5_embed.py once via point_stat and the obs block lines will access this dataframe according to what is in the config file:
obs->field->name

is my thinking correct?

**Reply by DanielAdriaansen** [reply]
> When running point_stat with observation python embedding and multiple observations lines in the obs block in the config file, will point_stat attempt to execute the python embedding once or as many times as there are lines in the obs block?

I believe it does it one time for each time. So if you change to a new valid time, the Python embedding (HDF5 read) will occur again for that new time, but does not run each time a new line is encountered in the obs block for the current time.

> Since the observation embedding is performed in the invocation of point_stat at the command line and since my observation processing python routine is only able to read in one variable into memory at a time, does this mean that I will need to run point_stat separately for each variable and hence point_stat will be reading in the same raob file once per variable?

If you are memory limited from a computing hardware perspective, then yes. But if not, I would suggest modifying your Python routine to read in all observations relevant for all lines in the obs block in the Configuration file for Point Stat, especially if you have to read the entire HDF5 file to memory.

> it seems like I should be able to modify my dataframe to have all the variable types in column 'var', and just run test_h5_embed.py once via point_stat and the obs block lines will access this dataframe according to what is in the config file:
> obs->field->name

That's right. Just give the "var" column in your DataFrame the name in each `name=` entry in the `obs` block and it should be able to find matches that way.

**Reply by justintsu** [reply]
thanks, I modified my script after my last post and deployed point_stat for all valid times/variables/pressure levels overnight.  It looks like it is taking 6-7 hours to run each point_stat for a valid time, which is a little less than ideal.  I've posted the cnt and config file for one of the valid times just because I am noticing some very large biases in the temperature, particularly at high and low elevations.  I suspect perhaps a unit error so I am checking my code 
[tsu.zip](https://github.com/dtcenter/METplus/files/13230207/tsu.zip)

**Reply by DanielAdriaansen** [reply, marked as answer]
Keep in mind that while the observations are only read one time with Python for each time, your COAMPS Python script is called for each entry in the `fcst` block. So you may consider doing some profiling of how long that script takes to run. Is it global data? What is the nx/ny of the grid?

I'd suggest reducing your config file to a single time, single subset of variables at a single level. It's possible some variables are taking longer than others for some reason, but without some systematic profiling it's hard to know where the bottlenecks are, if there are any.

One question I'd ask is whether it is necessary for you to be setting `MET_PYTHON_EXE`? If you set this, MET will write out and read back in any data that is obtained by Python. It is only recommended to use this if there are Python packages not available (or not possible to be installed) into the Python version that was used to compile the MET software against.

If you are able to use the Python MET was compiled with to run your Python scripts, data is passed in memory from Python to MET and no write/read is performed. 

For `read_COAMPS.py` I see:
```
import os
import sys
import re
import numpy as np
import datetime as dt
from IPython import embed as shell
```
I only see `shell()` used in one place, and it could probably be re-worked to not use that. All of the other imports here should be available in the Python3 standard library.

For your obs script, I see:
```
import h5py
from collections import OrderedDict
import pandas as pd

from IPython import embed as shell
import sys

import numpy as np
import math

#FOR PROGRESS BAR, COMMENT OUT IF YOU DONT HAVE MODULE
from tqdm import tqdm
```
So seems like you could get by without `tqdm`, and I don't see `shell` used anywhere here so that line can be removed. The only one that's special here is `h5py`. So, is it possible for you to check if the Python used when compiling MET has `h5py` installed?

If not, then you will need to set `MET_PYTHON_EXE`. Unfortunately there's no way to tell MET right now to use the compile time Python for FCST and the user's Python (at `MET_PYTHON_EXE`) for OBS so it will use `MET_PYTHON_EXE` for both which will force the writing/reading of the FCST data too.

Thus, it's probably advantageous given the scale of your task to see if you can switch to the compile time Python. Let me know if you have additional questions.

The mpr file should show the obs/fcst pair values to assist with units debugging, but I see that's off. You could turn it on and then run a single time for only 1 level for each variable if that would help.

**Reply by DanielAdriaansen** [reply]
Hi @justintsu, just following up on this. If you feel that we've resolved your issue, could you mark a response as the answer? Otherwise let us know if you need anything else. Thanks!
